include/openthread/thread.h

```diff
@@ -505,6 +505,38 @@ const char *otThreadGetDomainName(otInstance *aInstance);
  */
 otError otThreadSetDomainName(otInstance *aInstance, const char *aDomainName);
 
+/**
+ * Set/Clear the Interface Identifier manually specified for the Thread Domain Unicast Address.
+ *
+ * This function is only available since Thread 1.2 when `OPENTHREAD_CONFIG_DUA_ENABLE` is enabled.
+ *
+ * @param[in]  aInstance   A pointer to an OpenThread instance.
+ * @param[in]  aIid        A pointer to the Interface Identifier to set or NULL to clear.
+ *
+ * @retval OT_ERROR_NONE           Successfully set/cleared the Interface Identifier.
+ * @retval OT_ERROR_ALREADY        The same Interface Identifier was set already.
```

Comment:
I can still see `OT_ERROR_ALREADY/OT_ERROR_NOT_FOUND ` usage in public API like
https://github.com/openthread/openthread/blob/752ad977db187ab9357621ffd7e6247ef075f481/include/openthread/ip6.h#L226-L246 

Any principle in avoiding `OT_ERROR_ALREADY` 
 and `OT_ERROR_NOT_FOUND` to follow? @jwhui @simonlingoogle

Reply:
In many cases, I agree that we should avoid `OT_ERROR_ALREADY` and `OT_ERROR_NOT_FOUND`. Can we think of any important action that the caller would take on these errors? If not, we should probably return `void`.

Reply:
For API User, in some extent, I think `OT_ERROR_ALREADY` or `OT_ERROR_NOT_FOUND` may indicate some logic issue/state in user's application during development or debug. 

It might be not easy to decide whether or not such error code is important for the user for all scenarios. However returning such error code could give the user the flexibility to ignore the error code he/she doesn't care or take care when required. 

Thoughts?

Reply:
> For API User, in some extent, I think OT_ERROR_ALREADY or OT_ERROR_NOT_FOUND may indicate some logic issue/state in user's application during development or debug.

`OT_ASSERT` is the right way to catch logic issues? But I think it would be too strict. 

> However returning such error code could give the user the flexibility to ignore the error code he/she doesn't care or take care when required.

I think the whole point of removing OT_ERROR_ALREADY and OT_ERROR_NOT_FOUND is to make the API simpler and errors easier to handle. Users can still check IID by themselves if they need to catch logical errors as below:
```c++
    OT_ASSERT(otThreadGetFixedDuaInterfaceIdentifier(instance) == NULL);
    otThreadSetFixedDuaInterfaceIdentifier(instance, fixedIid);
```

Reply:
I also feel that returning `ALREADY`/`NOT_FOUND` errors in these cases is not that helpful and  actually generally requires more logic/code for driver/user side to correctly ignore such errors (I know we had to deal with such errors in `wpantund`/NCP and RTOS drivers for managing netif addresses). I would favor keeping it simple and returning `OT_ERROR_NONE` for such cases.

In case OT API users do need to detect if an address is already present or not, we can add API to get the address (or check an address). I think such API model would help make the intention of OT App developer (users of OT API)  more clear  compared to situation that they use the returned `otError`.